In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
from sklearn.model_selection import cross_validate
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn import datasets
from sklearn.metrics import accuracy_score
from sklearn import tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.linear_model import LogisticRegression

Build a classifier to predict the Target column using the other 20 features. 

In [ ]:
df=pd.read_csv('/content/drive/MyDrive/Data 599/classify_imb.csv')
df.head(1)

When you split your data into training and test sets, set random_state = 5. (2 points)

In [ ]:
X=df.drop('Target',axis=1)
y=df['Target']

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3, random_state=5)

Try a random forest (use random_state = 5 and 400 trees), a boosting classifier (use n_estimators = 300, learning rate =0.1, and max_depth =3), and a logistic regression classifier. (9 points: 3 for each classifier)

In [ ]:
#random forest
rf = RandomForestClassifier(n_estimators=400,max_features='sqrt',random_state=5).fit(X_train,y_train)

In [ ]:
accuracy_score(rf.predict(X_test),y_test)

0.7766666666666666

In [ ]:
pd.crosstab(y_test,rf.predict(X_test),rownames=['True'], colnames=['Predicted'])

Predicted    0   1
True              
0          190  10
1           57  43

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
print(classification_report(y_test,rf.predict(X_test)))

              precision    recall  f1-score   support

           0       0.77      0.95      0.85       200
           1       0.81      0.43      0.56       100

    accuracy                           0.78       300
   macro avg       0.79      0.69      0.71       300
weighted avg       0.78      0.78      0.75       300



In [ ]:
#boosting
bclf = GradientBoostingClassifier(n_estimators=300,max_depth=3,learning_rate=0.1,random_state=5).fit(X_train,y_train)

In [ ]:
accuracy_score(bclf.predict(X_test),y_test)

0.7566666666666667

In [ ]:
pd.crosstab(y_test,bclf.predict(X_test),rownames=['True'], colnames=['Predicted'])

Predicted    0   1
True              
0          181  19
1           54  46

In [ ]:
print(classification_report(y_test,bclf.predict(X_test)))

              precision    recall  f1-score   support

           0       0.77      0.91      0.83       200
           1       0.71      0.46      0.56       100

    accuracy                           0.76       300
   macro avg       0.74      0.68      0.69       300
weighted avg       0.75      0.76      0.74       300



In [ ]:
#log reg classifier

In [ ]:
logreg=LogisticRegression().fit(X_train,y_train)

In [ ]:
logreg.score(X_test,y_test)

0.6833333333333333

In [ ]:
print(classification_report(y_test,logreg.predict(X_test)))

              precision    recall  f1-score   support

           0       0.69      0.97      0.80       200
           1       0.65      0.11      0.19       100

    accuracy                           0.68       300
   macro avg       0.67      0.54      0.50       300
weighted avg       0.67      0.68      0.60       300



Which is best? How can you tell? (6 points)

Judging by the f1-score random forest and boosting classifiers did the best at 56% each, with random forest having a slightly stronger accuracy at 77.6%. Therefore, overall, random forest is the best model of the three. The logistic regression di the worst on f1 and accuracy. 

Comment on the precision, recall, and accuracy of your model. (3 points)

Precision, recall, and accuracy are all important metrics depending on what your model is measuring. In regards to our model, since accuracy and f1-score are almost the same in random forest and boosting classification, looking at precision and recall may be most beneficial for which metrics we want our model to capture. If we want to look at how well the model captures observations in Class 1, recall is the important metric to refer to. Boosting outperforms random forest in this area. If we want to know how many predicted 1's were actually 1 then we focus on precision, and random forest beats out boosting in this metric. Random forest beats out boosting slightly in accuracy.